lib/uksched/thread.c

```diff
@@ -101,107 +87,747 @@ extern const struct uk_thread_inittab_entry _uk_thread_inittab_end;
+				     ? 'T' : '-',
+				    (child->flags & UK_THREADF_ECTX)
+				     ? 'E' : '-',
+				    (child->flags & UK_THREADF_UKTLS)
+				     ? 'T' : '-');
+			continue;
 		}
+
+		uk_pr_debug("uk_thread %p (%s) term: Call termination %p()...\n",
+			    child, child->name
+			    ? child->name : "<unnamed>",
+			    *itr->term);
+		term_args.term = itr->term;
+		ukplat_tlsp_exec(child->uktlsp, _inittab_call_term, &term_args);
 	}
 
-	memset(thread, 0, sizeof(*thread));
-	thread->ectx = ectx;
-	thread->name = name;
-	thread->stack = stack;
-	thread->tls = tls;
-	thread->tlsp = (tls == NULL) ? 0 : ukarch_tls_tlsp(tls);
-	thread->entry = function;
-	thread->arg = arg;
-
-	/* Not runnable, not exited, not sleeping */
-	thread->flags = 0;
-	thread->wakeup_time = 0LL;
-	thread->detached = false;
-	uk_waitq_init(&thread->waiting_threads);
-	thread->sched = NULL;
-	thread->prv = NULL;
+	return ret;
+}
+
+static void _uk_thread_struct_init(struct uk_thread *t,
+				   uintptr_t tlsp,
+				   bool is_uktls,
+				   struct ukarch_ectx *ectx,
+				   const char *name,
+				   void *priv,
+				   uk_thread_dtor_t dtor)
+{
+	/* TLS pointer required if is_uktls is set */
+	UK_ASSERT(!is_uktls || tlsp);
+
+	memset(t, 0x0, sizeof(*t));
+
+	t->ectx = ectx;
+	t->tlsp = tlsp;
+	t->name = name;
+	t->priv = priv;
+	t->dtor = dtor;
+
+	if (tlsp && is_uktls)
+		t->flags |= UK_THREADF_UKTLS;
+	if (ectx) {
+		ukarch_ectx_init(t->ectx);
+		t->flags |= UK_THREADF_ECTX;
+	}
 
+	uk_waitq_init(&t->waiting_threads);
+
+#ifdef CONFIG_LIBNEWLIBC
 	/* TODO: Move newlibc reent initialization to newlib as
-	 *       thread initialization function
+	 *       thread initialization function and use TLS
 	 */
-#ifdef CONFIG_LIBNEWLIBC
-	reent_init(&thread->reent);
+	reent_init(&t->reent);
 #endif
 
-	/* Iterate over registered thread initialization functions */
-	uk_thread_inittab_foreach(itr) {
-		if (unlikely(!itr->init))
-			continue;
+	uk_pr_debug("uk_thread %p (%s): ctx:%p, ectx:%p, tlsp:%p\n",
+		    t, t->name ? t->name : "<unnamed>",
+		    &t->ctx, t->ectx, (void *) t->tlsp);
+}
 
-		uk_pr_debug("New thread %p: Call thread initialization function %p...\n",
-			    thread, *itr->init);
-		ret = (itr->init)(thread);
-		if (ret < 0)
-			goto err_fini;
+int uk_thread_init_bare(struct uk_thread *t,
+			uintptr_t ip,
+			uintptr_t sp,
+			uintptr_t tlsp,
+			bool is_uktls,
+			struct ukarch_ectx *ectx,
+			const char *name,
+			void *priv,
+			uk_thread_dtor_t dtor)
+{
+	UK_ASSERT(t);
+	UK_ASSERT(t != uk_thread_current());
+
+	_uk_thread_struct_init(t, tlsp, is_uktls, ectx, name, priv, dtor);
+	ukarch_ctx_init_bare(&t->ctx, sp, ip);
+
+	if (ip)
+		t->flags |= UK_THREADF_RUNNABLE;
+
+	return _uk_thread_call_inittab(t);
+}
+
+int uk_thread_init_bare_fn0(struct uk_thread *t,
+			    uk_thread_fn0_t fn,
+			    uintptr_t sp,
+			    uintptr_t tlsp,
+			    bool is_uktls,
+			    struct ukarch_ectx *ectx,
+			    const char *name,
+			    void *priv,
+			    uk_thread_dtor_t dtor)
```

Comment:
Yes, this is true. I do this normally, too. In fact, I have a kind of exception only with functions with a lot of arguments (don't ask me for the threshold that I am using 😉 ). I noticed that pressing as much as arguments possible into a few lines does not contribute to the readability. Here, we actually have a rather unbalanced length of arguments that I find tedious to read. I am prefer keeping it this way because of this, what do you think?

Reply:
Yeah, I also like it like this more. Just wanted to know if there's a convention for this, for the future.

I'm marking as resolved.